File: backend/packages/common/src/password_policy.py
"""Password-strength policy — server-side source of truth.

Mirrors frontend/trader/src/lib/passwordPolicy.ts. If you change a
rule here, update the frontend too — the frontend gives the user
real-time feedback before submit, but the backend is what actually
rejects weak passwords. Either layer alone is insufficient: client
checks can be bypassed, server-only checks frustrate the user with
"submit → fail → retry" loops.

Rules:
  1. Minimum 8 characters (hard floor)
  2. At least 3 of these 4 character classes: lower / upper / digit / symbol
  3. Not in the COMMON_WEAK_PASSWORDS blocklist
  4. Not a substring match against user-supplied disallow strings
     (typically the email local-part + first / last name)

A password that satisfies all four is accepted. Anything else raises
PasswordTooWeak with a short message the caller can return to the
client.
"""
from __future__ import annotations

import re
from typing import Iterable


class PasswordTooWeak(ValueError):
    """Raised when a candidate password fails policy. Use the .reason
    attribute (first failing rule) as the user-facing detail message."""

    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


# Top weak passwords — kept short on purpose (mirrors the frontend list).
# The Have-I-Been-Pwned corpus has millions; this set covers the ~70% of
# real brute-force attempts. Add to it freely when you spot a leak.
COMMON_WEAK_PASSWORDS: frozenset[str] = frozenset({
    "12345678", "123456789", "1234567890", "qwerty123", "password", "password1",
    "password123", "qwertyuiop", "asdfghjkl", "11111111", "00000000", "88888888",
    "abcdefgh", "abc12345", "iloveyou", "iloveyou1", "sunshine", "princess",
    "admin123", "admin1234", "letmein", "letmein123", "welcome1", "welcome123",
    "monkey123", "football", "baseball", "qwerty12", "asdf1234", "zxcvbnm",
    "qazwsxedc", "1q2w3e4r", "1qaz2wsx", "q1w2e3r4", "master123", "shadow123",
    "dragon123", "trustno1", "startrek", "starwars", "superman", "batman123",
    "pakistan", "india123", "hello123", "changeme", "demo1234", "test1234",
    "guest123", "user1234", "root1234", "trustx123", "trustx2025", "trustx2026",
})

_LOWER_RE  = re.compile(r"[a-z]")
_UPPER_RE  = re.compile(r"[A-Z]")
_DIGIT_RE  = re.compile(r"\d")
_SYMBOL_RE = re.compile(r"[^A-Za-z0-9]")

MIN_LENGTH = 8
MIN_CHAR_CLASSES = 3   # of {lower, upper, digit, symbol}


def validate_password(password: str, disallow: Iterable[str] = ()) -> None:
    """Raise `PasswordTooWeak` if the password fails policy; return None
    if it passes. The `disallow` iterable seeds substring-match checks
    (typically the user's email local-part, first name, last name) so
    callers can prevent "name1234" / "email@1234" style passwords.

    The check is intentionally case-insensitive on the substring search.
    """
    pw = password or ""

    if len(pw) < MIN_LENGTH:
        raise PasswordTooWeak(f"Password must be at least {MIN_LENGTH} characters")

    if pw.lower() in COMMON_WEAK_PASSWORDS:
        raise PasswordTooWeak("This password is too common — pick something unique")

    class_count = sum([
        bool(_LOWER_RE.search(pw)),
        bool(_UPPER_RE.search(pw)),
        bool(_DIGIT_RE.search(pw)),
        bool(_SYMBOL_RE.search(pw)),
    ])
    if class_count < MIN_CHAR_CLASSES:
        raise PasswordTooWeak(
            "Use a mix of upper-case, lower-case, numbers and symbols "
            f"(at least {MIN_CHAR_CLASSES} of these)"
        )

    pw_low = pw.lower()
    for needle in disallow:
        n = (needle or "").lower().strip()
        if len(n) >= 3 and n in pw_low:
            raise PasswordTooWeak("Don't reuse your email or name in your password")

File: frontend/trader/src/lib/passwordPolicy.ts
/**
 * Shared password-strength policy for every signup / reset / change-password
 * surface in the trader app.
 *
 * Mirrors backend/packages/common/src/password_policy.py — if you change
 * thresholds here, change them there too (server is the source of truth;
 * the frontend just gives the user real-time feedback before submit).
 *
 * Rules:
 *   1. Minimum 8 characters (hard floor)
 *   2. At least 3 of these 4 character classes: lower, upper, digit, symbol
 *   3. Not a known weak / breached password (small inline blocklist of
 *      the ~60 most-tried passwords — covers >70% of real-world brute force)
 *   4. Not the user's own email local-part (passed in as `disallow`)
 *
 * A password that satisfies #1-#4 scores at least "Good" (3). Anything
 * below "Good" is rejected at submit time.
 */

/** Top weak passwords + obvious sequence patterns. Lower-cased on check.
 *  Kept short on purpose — full breach corpus belongs on the server.   */
const COMMON_WEAK_PASSWORDS = new Set([
  "12345678",
  "123456789",
  "1234567890",
  "qwerty123",
  "password",
  "password1",
  "password123",
  "qwertyuiop",
  "asdfghjkl",
  "11111111",
  "00000000",
  "88888888",
  "abcdefgh",
  "abc12345",
  "iloveyou",
  "iloveyou1",
  "sunshine",
  "princess",
  "admin123",
  "admin1234",
  "letmein",
  "letmein123",
  "welcome1",
  "welcome123",
  "monkey123",
  "football",
  "baseball",
  "qwerty12",
  "asdf1234",
  "zxcvbnm",
  "qazwsxedc",
  "1q2w3e4r",
  "1qaz2wsx",
  "q1w2e3r4",
  "master123",
  "shadow123",
  "dragon123",
  "trustno1",
  "startrek",
  "starwars",
  "superman",
  "batman123",
  "pakistan",
  "india123",
  "hello123",
  "changeme",
  "demo1234",
  "test1234",
  "guest123",
  "user1234",
  "root1234",
  "trustx123",
  "trustx2025",
  "trustx2026",
]);

export type Strength = 0 | 1 | 2 | 3 | 4;

export interface PasswordCheck {
  /** 0=empty, 1=very weak, 2=weak, 3=good, 4=strong */
  score: Strength;
  /** Human label for the score; e.g. "Weak", "Good". */
  label: string;
  /** Tailwind/hex colour for the strength bar at this score. */
  color: string;
  /** Per-rule pass/fail breakdown — feed straight to a checklist UI. */
  checks: {
    length: boolean;
    lowercase: boolean;
    uppercase: boolean;
    digit: boolean;
    symbol: boolean;
    notCommon: boolean;
    notLikeEmail: boolean;
  };
  /** Plain-language list of what's still missing — at most a few items. */
  issues: string[];
  /** True iff the password is acceptable for submit (score >= 3). */
  acceptable: boolean;
}

const LABELS: Record<Strength, string> = {
  0: "",
  1: "Very weak",
  2: "Weak",
  3: "Good",
  4: "Strong",
};

const COLORS: Record<Strength, string> = {
  0: "#9ca3af",
  1: "#ef4444", // red
  2: "#f59e0b", // amber
  3: "#22c55e", // green
  4: "#035eeb", // brand green
};

/**
 * Score a candidate password. `disallow` is an optional list of substrings
 * we don't want to see inside the password (typically the user's email
 * local-part and first name).
 */
export function scorePassword(
  password: string,
  disallow: string[] = [],
): PasswordCheck {
  const pw = password ?? "";
  if (pw.length === 0) {
    return {
      score: 0,
      label: "",
      color: COLORS[0],
      checks: {
        length: false,
        lowercase: false,
        uppercase: false,
        digit: false,
        symbol: false,
        notCommon: true,
        notLikeEmail: true,
      },
      issues: [],
      acceptable: false,
    };
  }

  const lowered = pw.toLowerCase();
  const checks = {
    length: pw.length >= 8,
    lowercase: /[a-z]/.test(pw),
    uppercase: /[A-Z]/.test(pw),
    digit: /\d/.test(pw),
    symbol: /[^A-Za-z0-9]/.test(pw),
    notCommon: !COMMON_WEAK_PASSWORDS.has(lowered),
    notLikeEmail: disallow
      .map((s) => (s || "").toLowerCase().trim())
      .filter((s) => s.length >= 3)
      .every((s) => !lowered.includes(s)),
  };

  // Quick deal-breakers: anything failing length / common / email match
  // can't score above Weak no matter what else is true. Catches
  // "Password1@" (passes char classes) being banned because it's common.
  const classCount = [
    checks.lowercase,
    checks.uppercase,
    checks.digit,
    checks.symbol,
  ].filter(Boolean).length;

  let score: Strength;
  const issues: string[] = [];
  if (!checks.length) issues.push("At least 8 characters");
  if (!checks.notCommon) issues.push("Too common — pick something unique");
  if (!checks.notLikeEmail) issues.push("Don't reuse your email or name");
  if (classCount < 3) {
    const missing: string[] = [];
    if (!checks.lowercase) missing.push("a lower-case letter");
    if (!checks.uppercase) missing.push("an upper-case letter");
    if (!checks.digit) missing.push("a number");
    if (!checks.symbol) missing.push("a symbol");
    issues.push(`Add ${missing.slice(0, 4 - classCount).join(" or ")}`);
  }

  // Base score on character classes + length bonus, then cap by rule failures.
  let raw = classCount;
  if (pw.length >= 12) raw += 1;
  if (pw.length >= 16) raw += 1;
  // Map raw (0-6) → display score (1-4).
  if (raw <= 1) score = 1;
  else if (raw === 2) score = 2;
  else if (raw <= 4) score = 3;
  else score = 4;

  // Hard caps for rule violations.
  if (!checks.length) score = 1;
  if (!checks.notCommon) score = Math.min(score, 1) as Strength;
  if (!checks.notLikeEmail) score = Math.min(score, 2) as Strength;
  if (classCount < 3) score = Math.min(score, 2) as Strength;

  const acceptable =
    score >= 3 &&
    checks.length &&
    checks.notCommon &&
    checks.notLikeEmail &&
    classCount >= 3;

  return {
    score,
    label: LABELS[score],
    color: COLORS[score],
    checks,
    issues,
    acceptable,
  };
}

/** Single requirement spec — fed to the checklist UI on signup / reset. */
export interface Requirement {
  id: keyof PasswordCheck["checks"];
  label: string;
}

export const PASSWORD_REQUIREMENTS: Requirement[] = [
  { id: "length", label: "8+ characters" },
  { id: "uppercase", label: "An upper-case letter (A–Z)" },
  { id: "lowercase", label: "A lower-case letter (a–z)" },
  { id: "digit", label: "A number (0–9)" },
  { id: "symbol", label: "A symbol (!@#$…)" },
  { id: "notCommon", label: "Not a common password" },
];
